consensus/src/experimental/buffer_manager.rs

```diff
@@ -254,6 +184,112 @@ impl StateManager {
         Ok(())
     }
 
+    /// this function updates the buffer according to the response from the execution phase
+    /// it also initiates a request to the signing phase.
+    async fn update_buffer_execution_resp(
+        &mut self,
+        execution_resp: ExecutionResponse,
+    ) -> anyhow::Result<()> {
+        let ExecutionResponse { inner } = execution_resp;
+
+        let executed_blocks = inner?;
+
+        let current_cursor = get_next(&self.execution_root); // must be Some
+
+        if current_cursor.is_some() {
+            // update buffer
+            let buffer_item = take_elem(&current_cursor);
+            if let BufferItem::Ordered(ordered_box) = &buffer_item {
+                if ordered_box.ordered_blocks.first().unwrap().id()
+                    != executed_blocks.first().unwrap().id()
+                {
+                    // an sync req happened before the response
+                    // we do nothing except putting the item back
+                    // the process_execution_resp function will retry the next ordered batch
+                    set_elem(&current_cursor, buffer_item);
+                    return Ok(());
+                }
+                for (ordered_block, executed_block) in
+                    zip(ordered_box.ordered_blocks.iter(), executed_blocks.iter())
+                {
+                    assert_eq!(ordered_block.id(), executed_block.id())
```

Comment:
I thought we already do this check inside advance_to_executed?

Reply:
resolved